### PR TITLE
[mysql]: Fix limited songs query after #8993

### PR DESCRIPTION
**Pull request by MilhouseVH**
See [comment](https://github.com/xbmc/xbmc/pull/8993#issuecomment-178763842) in #8993.

### CODE REVIEW

File: xbmc/music/MusicDatabase.cpp [hunk 1]
```diff
@@ -4079,7 +4079,7 @@ bool CMusicDatabase::GetSongsFullByWhere(const std::string &baseDir, const Filte
         strSQL += strSQLExtra;
       else
         //Apply where clause and limits to songview, then join as mutiple records in result set per song
-        strSQL += " WHERE songview.idsong in (SELECT idsong FROM songview " + strSQLExtra + ")";
+        strSQL += " WHERE songview.idsong IN (SELECT idsong FROM (SELECT idsong FROM songview " + strSQLExtra + ") as t)";
```
Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: Very minor, but why alias it with "t" when you never use that alias?

Comment (MilhouseVH) [inline xbmc/music/MusicDatabase.cpp]: No particular reason, pretty much cut & pasted from somewhere else where by calling it `t` they're indicating it's a temp table. Will drop `as t`

Comment (MilhouseVH) [inline xbmc/music/MusicDatabase.cpp]: Or maybe I won't drop it... without `as t`, MySQL fails with:
```
Error Code: 1248. Every derived table must have its own alias
```
Will rename `t` -> `temp` - any objections?

Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: nope, everything is better than "t" ;)

**Comment by MilhouseVH** [comment]
Let me know if you need a backport of this for #9005

**Comment by MartijnKaijser** [comment]
if this is in Jarvis backport should be needed yes. plan is for last rc3 tomorrow morning

**Comment by MilhouseVH** [comment]
@MartijnKaijser : Done, #9036.

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
Thanks for this @MilhouseVH . I must admit I totally forgot about MySQL testing, sorry.

What the query is trying to achieve is returning records for a limited number of songs, but the query itself is on a join to song_artist so can have multiple records for every song. The where clause built can already have a number of "idsong in..." clauses to select songs, so the resulting query could look rather odd.

The only other approach I can thing of is to count the number of artists that the required number of otherwise filtered songs has and them apply that as the limit on the join query. Any other suggestions welcome.

**Comment by Razzeee** [comment]
Can't we just limit the top query? As the status quo also picks a random ID it shouldn't matter really?
So we need to split the Limit from strSQLExtra and apply strSQLExtra normally and limit the one above.
    strSQL += " WHERE songview.idsong IN (SELECT idsong FROM songview " + strSQLExtra + ") " + strLimit;

Something like that.

**Comment by phate89** [comment]
@DaveTBlake 
It's not possible to use subqueries? For example
```
"SELECT %s FROM (SELECT idsong FROM songview " + strSQLExtra + ") AS sv LEFT JOIN songartistview on songartistview.idsong = sv.idsong 
```
In mysql nesting several queries has terrible performance..

**Comment by MilhouseVH** [comment]
@Razzeee - it's possible, although I'm just fixing the original query so that it works on MySQL. I don't know if what you propose has unforeseen side-effects which is why I didn't propose it (I did test it while fixing the original query, it produces the same result, but I've only got a single test case to work with - LIMIT 1 - and I'm still not 100% sure what this query is meant to achieve or why it is written the way it is so probably best for @DaveTBlake to respond).

**Comment by DaveTBlake** [comment]
@Razzeee , @MilhouseVH I'm away at the moment so can't test and don't really have time to think about t until the weekend. But what I think @Razzeee you suggest is where I started: limit on outer clause applies the limit to song join aong_artist, so rather than limit the number of songs returned it limits the number of artists and songs. So say you want 1 song (odd use Millhouse?) but it has 3 artists, limit 1 would get to just the song and the first artist, the other artists would be missed.

As for what this query is trying to do: return song data including the artists that are in song_artists (song.strArtists is now just a desc field not a denormalisation of the song_artist link), that satisfy filter criteria. The filter could be any combination of playlist rules, or JSON parameters etc. But also a limit on the number of songs returned can be applied - mostly so that JSON can return song data in chunks.

**Comment by DaveTBlake** [comment]
@phate89 said (and I don't know why I got Git email but can't see comment here?)
>@DaveTBlake
It's not possible to use subqueries? For example

>"SELECT %s FROM (SELECT idsong FROM songview " + strSQLExtra + ") AS sv LEFT JOIN songartistview on songartistview.idsong = sv.idsong 

>In mysql nesting several queries has terrible performance.

Interesting suggestion, thanks Phate. Need to test if can apply the limit or may still have the problem with MySQL. But otherwise could make a better query generally.

**Comment by MilhouseVH** [comment]
> odd use Millhouse?

It's a method (trick?) to quickly determine the total number of items, in this case Songs, by querying for all items but limiting to one item - JSON will return the total number of songs in the `limits` result: 

```
2016-02-04 08:46:20.367163:MainThread: libStats.JSON SOCKET REQUEST: [{"jsonrpc": "2.0", "params": {"limits": {"start": 0, "end": 1}}, "method": "AudioLibrary.GetSongs", "id": "libStats"}]
2016-02-04 08:46:20.448652:MainThread: libStats.BUFFER RECEIVED (len 128)
2016-02-04 08:46:20.448713:MainThread: libStats.PARSING JSON DATA: {"id":"libStats","jsonrpc":"2.0","result":{"limits":{"end":1,"start":0,"total":7388},"songs":[{"label":"Avenue A","songid":1}]}}
```

In this case, 7388 songs. I use this when calling `texturecache.py stats`.

**Comment by DaveTBlake** [comment]
Thanks for explaining use @MilhouseVH , would a count method in JSON be preferable?

**Comment by MilhouseVH** [comment]
> would a count method in JSON be preferable?

From a JSON perspective the current method is fine (IMHO), although I'm not sure how the `total` figure is being determined - hopefully a `count()` is happening somewhere rather than reading all rows and throwing them away. Maybe @Montellese can shed some light on it.

**Comment by MilhouseVH** [comment]
Just a quick update on the question of `count()`.

The following JSON query:
```
{"jsonrpc": "2.0", "params": {"limits": {"start": 0, "end": 1}}, "method": "AudioLibrary.GetSongs", "id": "libStats"}
```

produces the following MySQL queries:
```
                1390208 Query   SELECT COUNT(1) FROM songview
                1390208 Query   SELECT songview.*, songartistview.*  FROM songview LEFT JOIN songartistview on songartistview.idsong = songview.idsong  WHERE songview.idsong IN (SELECT idsong FROM (SELECT idsong FROM songview  LIMIT 1) as temp)
                1390208 Query   select * from path where strPath='nfs://192.168.0.3/mnt/share/data/Music/MP3/ABC/Alphabet City/'
                1390208 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - Avenue A.mp3'
```
so it looks like `total` is the result of a `count()`. The second `SELECT` (the object of this PR) returns just a single row.

The same JSON query, with `end: 10`, produces the following MySQL queries - the repeated `cue` queries wouldn't be much fun for all 7388 items but that's possibly a PR for another day...

```
                1390221 Query   SELECT COUNT(1) FROM songview
                1390221 Query   SELECT songview.*, songartistview.*  FROM songview LEFT JOIN songartistview on songartistview.idsong = songview.idsong  WHERE songview.idsong IN (SELECT idsong FROM (SELECT idsong FROM songview  LIMIT 10) as temp)
                1390221 Query   select * from path where strPath='nfs://192.168.0.3/mnt/share/data/Music/MP3/ABC/Alphabet City/'
                1390221 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - Avenue A.mp3'
                1390221 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - When Smokey Sings.mp3'
                1390221 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - The Night You Murdered Love.mp3'
                1390221 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - Think Again.mp3'
                1390221 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - Rage And Then Regret.mp3'
                1390221 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - Ark-Angel.mp3'
                1390221 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - King Without A Crown.mp3'
                1390221 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - Bad Blood.mp3'
                1390221 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - Jealous Lover.mp3'
                1390221 Query   select strCuesheet from cue where idPath=3 AND strFileName='ABC - One Day.mp3'
```

**Comment by DaveTBlake** [comment]
Yes a count query is performed, and I guess that also querying a single song (and artists) is not much of a unneeded overhead. But the cuesheet querying is something I was unaware of, and certainly worth some investigation for efficiency improvements!! I'll have a look separately.

Meanwhile @MilhouseVH , matter in hand, do queries of the form

`SELECT sv.*, songartistview.* FROM (select * FROM songview LIMIT 1) AS sv LEFT JOIN songartistview on songartistview.idsong = sv.idsong`

work in MySQL? Unfortunately I don't have anyway to test that for myself at the moment. But this seems a better format than the subquery.

**Comment by MilhouseVH** [comment]
> Meanwhile @MilhouseVH , matter in hand, do queries of the form

> SELECT sv.*, songartistview.* FROM (select * FROM songview LIMIT 1) AS sv LEFT JOIN songartistview on songartistview.idsong = sv.idsong

> work in MySQL? Unfortunately I don't have anyway to test that for myself at the moment. But this seems a better format than the subquery.

Yes, that's working - it returns a single row ("Avenue A", same as the `LIMIT 1` query addressed by this PR). I'm happy for this PR to be closed if you want to continue with an updated query - mine was just a quick & dirty fix. :)

**Comment by phate89** [comment]
@DaveTBlake discard my suggestion.. That query it's actually worse in speed...

**Comment by DaveTBlake** [comment]
That is surprising @phate89, it certainly reads easier and you would think that limiting before join would work well. I'm waiting on a rebuild before I can get testing.

**Comment by MilhouseVH** [comment]
Closing, see #9081.

**Pull request closed without merging**